code/espurna/sensors/PZEM004TSensor.h

```diff
@@ -104,34 +153,99 @@ class PZEM004TSensor : public BaseSensor {
-            return description();
+            int dev = index / PZ_MAGNITUDE_COUNT;
+            char buffer[25];
+            snprintf(buffer, sizeof(buffer), "(%u/%s)", dev, getAddress(dev).c_str());
+            return description() + String(buffer);
         };
 
         // Address of the sensor (it could be the GPIO or I2C address)
         String address(unsigned char index) {
-            return _ip.toString();
+            int dev = index / PZ_MAGNITUDE_COUNT;
+            return _devices[dev].toString();
         }
 
         // Type for slot # index
         unsigned char type(unsigned char index) {
-            if (index == 0) return MAGNITUDE_CURRENT;
-            if (index == 1) return MAGNITUDE_VOLTAGE;
-            if (index == 2) return MAGNITUDE_POWER_ACTIVE;
-            if (index == 3) return MAGNITUDE_ENERGY;
+            int dev = index / PZ_MAGNITUDE_COUNT;
+            index = index - (dev * PZ_MAGNITUDE_COUNT);
+            if (index == PZ_MAGNITUDE_CURRENT_INDEX)      return MAGNITUDE_CURRENT;
+            if (index == PZ_MAGNITUDE_VOLTAGE_INDEX)      return MAGNITUDE_VOLTAGE;
+            if (index == PZ_MAGNITUDE_POWER_ACTIVE_INDEX) return MAGNITUDE_POWER_ACTIVE;
+            if (index == PZ_MAGNITUDE_ENERGY_INDEX)       return MAGNITUDE_ENERGY;
             return MAGNITUDE_NONE;
         }
 
         // Current value for slot # index
         double value(unsigned char index) {
+            int dev = index / PZ_MAGNITUDE_COUNT;
+            index = index - (dev * PZ_MAGNITUDE_COUNT);
             double response = 0;
-            if (index == 0) response = _pzem->current(_ip);
-            if (index == 1) response = _pzem->voltage(_ip);
-            if (index == 2) response = _pzem->power(_ip);
-            if (index == 3) response = _energy_offset + (_pzem->energy(_ip) * 3600);
+            if (index == PZ_MAGNITUDE_CURRENT_INDEX)      response = _readings[dev].current;
+            if (index == PZ_MAGNITUDE_VOLTAGE_INDEX)      response = _readings[dev].voltage;
+            if (index == PZ_MAGNITUDE_POWER_ACTIVE_INDEX) response = _readings[dev].power;
+            if (index == PZ_MAGNITUDE_ENERGY_INDEX)       response = (_readings[dev].energy * 3600) - _energy_offsets[dev];
             if (response < 0) response = 0;
             return response;
         }
 
+        // Post-read hook (usually to reset things)
+        void post() {
+            _error = SENSOR_ERROR_OK;
+        }
+
+        // Loop-like method, call it in your main loop
+        void tick() {
+            static unsigned char dev = 0;
+            static unsigned char magnitude = 0;
+            static unsigned long last_millis = 0;
+
+            if (_busy || millis() - last_millis < PZEM004T_READ_INTERVAL) return;
+
+            _busy = true;
+
+            // Clear buffer in case of late response(Timeout)
+            while(Serial.available() > 0) Serial.read();
+
+            float read;
+            float* readings_p;
```

Comment:
In the switch case, you will see that we read the sensor and store it in `read` and in the `readings_p` we store the address that this value will be stored in case it is not an error.

`read` is not error value, it _*can*_ store an error in case of the sensor did not respond.

Just to clarify, every time the method `void tick()` is invoked, it will read a magnitude from a device. It will iterate over all magnitudes over all devices. This is to avoid long readings, as the PZEM004T can take up to 1second to respond. The initial version I did some months ago I used to read the magnitude inside the `value` method which caused the watchdog to trigger. In this version, I do the readings over time and when asked I return the last value read.

I read the same magnitudes on all devices first, so the readings are close to each other, current, then voltage, then power then energy.

Reply:
Ok, nvm, I see it now. Had this comment stored since yesterday and submitted without double checking...